### PR TITLE
Latest import csv

### PR DESCRIPTION
**Pull request by timifasubaa**
This PR allows users to import CSV files to superset. It extends the work that had already been done  in (https://github.com/apache/incubator-superset/pull/2381) and also adds the new functionality of uploading to a hive datasource. All feedback to make the code and user experience better is welcome.

@mistercrunch

### CODE REVIEW

**Comment by xrmx** [comment]
I think the approach of a pandas datasource as in #3492 is more promising

File: superset/__init__.py [hunk 1]
```diff
@@ -156,6 +156,7 @@ def index(self):
 
 get_session = appbuilder.get_session
 results_backend = app.config.get("RESULTS_BACKEND")
+csv_upload_backend = app.config.get("CSV_UPLOAD_BACKEND")
```
Comment (mistercrunch) [inline superset/__init__.py]: unused?

File: superset/config.py [hunk 2]
```diff
@@ -289,11 +293,21 @@ class CeleryConfig(object):
...
 # An instantiated derivative of werkzeug.contrib.cache.BaseCache
 # if enabled, it can be used to store the results of long-running queries
 # in SQL Lab by using the "Run Async" button/feature
 RESULTS_BACKEND = None
 
+# An instantiated derivative of werkzeug.contrib.cache.BaseCache
+# if enabled, it can be used to store external Hive tables on the
+# S3 location specified in CSV_UPLOAD_BUCKET
+#from S3Cache.s3cache import S3Cache
+CSV_UPLOAD_BACKEND = None #if not CSV_TO_HIVE_UPLOAD_BUCKET else S3Cache(CSV_TO_HIVE_UPLOAD_BUCKET, "EXTERNAL_HIVE_TABLES")
```
Comment (mistercrunch) [inline superset/config.py]: ununsed?

File: superset/db_engine_specs.py [hunk 3]
```diff
@@ -74,6 +81,103 @@ def extra_table_metadata(cls, database, table_name, schema_name):
...
+                                 quotechar=quotechar,
+                                 escapechar=escapechar,
+                                 comment=comment,
+                                 encoding='utf-8',
+                                 error_bad_lines=error_bad_lines,
+                                 chunksize=chunksize,
+                                 iterator=True)
+        df = pandas.DataFrame()
+        df = pandas.concat(chunk for chunk in chunks)
+        print(df)
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: nope

File: superset/db_engine_specs.py [hunk 4]
```diff
@@ -74,6 +81,103 @@ def extra_table_metadata(cls, database, table_name, schema_name):
...
+        engine = create_engine(con, echo=False)
+
+        # Use Pandas to parse dataframe to database
+        df.to_sql(name=name, con=engine, schema=schema, if_exists=if_exists,
+                  index=index, index_label=index_label, chunksize=chunksize)
+
+
+    @staticmethod
+    def upload_csv(form):
+        print(results_backend)
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: let's not leave all these print statements around. Use `logging.info` if you want to log things

File: superset/db_engine_specs.py [hunk 5]
```diff
@@ -74,6 +81,103 @@ def extra_table_metadata(cls, database, table_name, schema_name):
...
+        def allowed_file(filename):
+            # Only allow specific file extensions as specified in the config
+            return '.' in filename and \
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: using index `[-1]` would allow for filenames with multiples dots in them, but better is to use the `os.path` module as in https://docs.python.org/2/library/os.path.html#os.path.splitext

File: superset/db_engine_specs.py [hunk 6]
```diff
@@ -74,6 +81,103 @@ def extra_table_metadata(cls, database, table_name, schema_name):
...
+                            parse_dates=form.parse_dates.data,
+                            infer_datetime_format=form.infer_datetime_format.data,
+                            dayfirst=form.dayfirst.data,
+                            thousands=form.thousands.data,
+                            decimal=form.decimal.data,
+                            quotechar=form.quotechar.data,
+                            escapechar=form.escapechar.data,
+                            comment=form.comment.data,
+                            error_bad_lines=form.error_bad_lines.data,
+                            chunksize=10000)
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: seems arbitrary, unclear where it belongs at this point though, maybe a database configuration element? Ok leaving as is for now though.

File: superset/db_engine_specs.py [hunk 7]
```diff
@@ -686,6 +790,41 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
...
+        table_name=form.name.data
+        file_name = form.csv_file.data.filename
+        print(results_backend)
+
+        #from superset import csv_upload_backend
+        #csv_upload_backend.set("k", "v")
+        #TODO(timifasubaa): replace the current approach with the results backend approach above.
+
+        bucket_path = config["HIVE_CSV_BUCKET_PATH"] 
+        s3_dir = "hive_csv/"+table_name+"/"
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: PEP8, lint your diff, also use `os.path.join`

File: superset/db_engine_specs.py [hunk 8]
```diff
@@ -686,6 +790,41 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
...
+
+        upload_path = config['UPLOAD_FOLDER'] + secure_filename(form.csv_file.data.filename)
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: `os.path.join`

File: superset/db_engine_specs.py [hunk 9]
```diff
@@ -686,6 +790,41 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
...
+        column_names = get_column_names(upload_path)
+        schema_definition = ", ".join([col_name + " STRING " for col_name in column_names])
+
+        import boto3
+        s3 = boto3.client('s3')
+        s3.upload_file(upload_path, 'airbnb-superset', s3_dir+file_name)
+        sql = "CREATE EXTERNAL TABLE " + str(table_name) + " ( " + str(schema_definition) + " ) " +\
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: use a triple quoted string with `.format()` for long string assembly as in:
```
s = """
    CREATE EXTERNAL TABLE {table}
   ...
""".format(**locals())
```

File: superset/db_engine_specs.py [hunk 10]
```diff
@@ -686,6 +790,41 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
...
+            "ROW FORMAT DELIMITED FIELDS TERMINATED BY ',' STORED AS TEXTFILE " + "LOCATION '" + str(bucket_path) + str(s3_dir) + "'"
+        try:
+            engine = create_engine(form.con.data)
+            engine.execute(sql)
+        except Exception as e:
+            print(e)
```
[the remark this answers is not in the document]
Comment (mistercrunch) [reply superset/db_engine_specs.py]: Also, that means all errors get silenced. Errors need to bubble up to users.

File: superset/forms.py [hunk 11]
```diff
@@ -0,0 +1,1367 @@
+"""Contains the logic to create cohesive forms on the explore view"""
```
Comment (mistercrunch) [inline superset/forms.py]: Does this huge file belong in your PR? Seems like an artifact from explore v1.

File: superset/translations/it/LC_MESSAGES/messages.po [hunk 12]
```diff
@@ -2789,8 +2789,8 @@ msgstr ""
 
 #: superset/connectors/sqla/models.py:386
 msgid ""
-"Datetime column not provided as part table configuration and is required "
-"by this type of chart"
+"Defines the origin where time buckets start, accepts natural dates as in "
```
Comment (mistercrunch) [inline superset/translations/it/LC_MESSAGES/messages.po]: Let's keep all of the translations out of this PR

File: superset/views/core.py [hunk 13]
```diff
@@ -12,13 +12,14 @@
 import re
 import time
 import traceback
-
+import os
+import pandas
```
Comment (mistercrunch) [inline superset/views/core.py]: unused?

Comment (mistercrunch) [reply superset/views/core.py]: btw pylint should tell you when you have unused imports

File: superset/views/core.py [hunk 14]
```diff
@@ -298,6 +304,129 @@ class DatabaseAsync(DatabaseView):
...
+        form.error_bad_lines.data = False
+        form.names.data = None
+        form.con.data = config['SQLALCHEMY_DATABASE_URI']
+        form.schema.data = None
+        form.if_exists.data = 'append'
+        form.index.data = None
+        form.index_label.data = None
+
+    def form_post(self, form):
+        # post process form
```
Comment (mistercrunch) [inline superset/views/core.py]: method name is self-explanatory, comment is confusing or doesn't add value

File: superset/views/core.py [hunk 15]
```diff
@@ -298,6 +304,129 @@ class DatabaseAsync(DatabaseView):
...
+
+        # Turn into list of strings
```
Comment (mistercrunch) [inline superset/views/core.py]: unrelated?

File: superset/views/core.py [hunk 16]
```diff
@@ -298,6 +304,129 @@ class DatabaseAsync(DatabaseView):
...
+            if csv_file and csv_file.filename:
+                filename = secure_filename(csv_file.filename)
+                csv_file.save(os.path.join(config['UPLOAD_FOLDER'],
+                                                 filename))
+                return filename
+
+
+        if form.names.data is not None:
+            form.names.data = form.names.data.split(",")
+        else:
```
Comment (mistercrunch) [inline superset/views/core.py]: elif?

File: superset/views/core.py [hunk 17]
```diff
@@ -298,6 +304,129 @@ class DatabaseAsync(DatabaseView):
...
+            if form.header.data is None:
+                form.header.data = 0
+
+
+        datasources = ConnectorRegistry.get_all_datasources(db.session)
+        filename = upload_file(form.csv_file.data)
+
+        # some of this stuff belongs in db_engine_spec
+        #print(form.data)
```
Comment (mistercrunch) [inline superset/views/core.py]: please clean up those artifacts, makes me feel like this PR isn't ready for review

File: superset/views/core.py [hunk 18]
```diff
@@ -298,6 +304,129 @@ class DatabaseAsync(DatabaseView):
...
+        database = db.session.query(models.Database).filter_by(sqlalchemy_uri=form.data.get('con')).one()
```
Comment (mistercrunch) [inline superset/views/core.py]: line too long, try to respect the <=80-90 PEP8 guideline, you can use js-like method chaining syntax between parenthesis in python as in:
```
database = ( 
    db.session.query(models.Database)
    .filter_by(sqlalchemy_uri=form.data.get('con'))
    .one()
)
```

File: superset/views/core.py [hunk 19]
```diff
@@ -298,6 +304,129 @@ class DatabaseAsync(DatabaseView):
...
+            chunk = flask.request.stream.read(chunk_size).decode('utf-8')
+            if len(chunk) == 0:
+                return
+            f.write(chunk)
+
+    #return send_from_directory(config['UPLOAD_FOLDER'],
+    #                           filename)
+
+class CsvToDatabaseView(SimpleFormView):
+    #DatasourceModelView,  add to ^ ?
```
Comment (mistercrunch) [inline superset/views/core.py]: clean up comment, add doctring https://www.python.org/dev/peps/pep-0257/

File: tests/testCSV.csv [hunk 20]
```diff
@@ -0,0 +1,2 @@
+Name,TestField
```
Comment (mistercrunch) [inline tests/testCSV.csv]: ? cleanup

**Review by mistercrunch — CHANGES_REQUESTED** [review]
PR doesn't seem ready for review, please add a [WiP] work in progress tag when that's the case.

Also it'd be great to reference the original PR in the PR description. Github will notify people tracking that PR as a result.

**Comment by timifasubaa** [comment]
Sorry. I should have made it clearer that it was WIP. I've now updated it. I wanted to get general feedback on the approach I took before cleaning up the code and you've done that. Thank you.

File: superset/db_engine_specs.py [hunk 21]
```diff
@@ -74,6 +84,115 @@ def extra_table_metadata(cls, database, table_name, schema_name):
         """Returns engine-specific table metadata"""
         return {}
 
+    @staticmethod
+    def csv_to_df(names, filepath_or_buffer, sep, header, index_col, squeeze,
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: Some linters will trigger on "too many params" on methods. This has lots of boilerplate code here. We can chat about how to improve this using a data structure and an unpack operator `**`, but it may be ok as is. For instance you could have a `pd_read_csv_kwargs` dict that you'd pass around, and ultimately call `pandas.read_csv(**pd_read_csv_kwargs)`

File: superset/db_engine_specs.py [hunk 22]
```diff
@@ -74,6 +84,115 @@ def extra_table_metadata(cls, database, table_name, schema_name):
...
+    def df_to_db(df, name, con, schema, if_exists, index,
+                 index_label, chunksize):
+        engine = create_engine(con, echo=False)
+        # Use Pandas to parse dataframe to database
+        df.to_sql(name=name, con=engine, schema=schema, if_exists=if_exists,
+                  index=index, index_label=index_label, chunksize=chunksize)
+
+        database = (
+            db.session
+                .query(type(query)) #Database
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: NIT: I think in most other places in the code base we line up the js-style method chaining with the first line. It's not PEP8 or anything, just within-codebase consistency.

File: superset/db_engine_specs.py [hunk 23]
```diff
@@ -74,6 +84,115 @@ def extra_table_metadata(cls, database, table_name, schema_name):
...
+            # Only allow specific file extensions as specified in the config
+            extension = os.path.splitext(filename)[1]
+            return extension and extension[1:] in config['ALLOWED_EXTENSIONS']
+
+        filename = secure_filename(form.csv_file.data.filename)
+        if not allowed_file(filename):
+            flash("Invalid file trype selected.", 'alert')
+            return False
+        df = BaseEngineSpec.csv_to_df(names=form.names.data,
+                            filepath_or_buffer=filename,
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: NIT: doesn't the linter trigger on this? `names` should be aligned with `filepath_or_buffer`, personally I'd break after the `(` and go with just 4 indents.

This is also a lot of boilerplate... There's a way to do this where the list of args is defined only once, in an array. We can chat about it.

File: superset/db_engine_specs.py [hunk 24]
```diff
@@ -74,6 +84,115 @@ def extra_table_metadata(cls, database, table_name, schema_name):
...
+                            dayfirst=form.dayfirst.data,
+                            thousands=form.thousands.data,
+                            decimal=form.decimal.data,
+                            quotechar=form.quotechar.data,
+                            escapechar=form.escapechar.data,
+                            comment=form.comment.data,
+                            error_bad_lines=form.error_bad_lines.data,
+                            chunksize=10000)
+
+        BaseEngineSpec.df_to_db(df=df,
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: NIT: break after `(` is your linter not triggering on this?

File: superset/db_engine_specs.py [hunk 25]
```diff
@@ -686,6 +805,50 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
         return BaseEngineSpec.fetch_result_sets(
             db, datasource_type, force=force)
 
+    @staticmethod
+    def upload_csv(form, table):
+        "uploads a cv file and creates a superset table in the right databse"
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: many (4?) typos and wrong capitalization

File: superset/db_engine_specs.py [hunk 26]
```diff
@@ -686,6 +805,50 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
...
+        "returns true if it was siccessful. False otherwise"
+        def get_column_names(filepath):
+            import csv
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: imports go at the top of the module, linter should trigger on this

File: superset/db_engine_specs.py [hunk 27]
```diff
@@ -686,6 +805,50 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
...
+        upload_prefix = config["CSV_TO_HIVE_UPLOAD_DIRECTORY"]
+        dest_path = os.path.join(table_name, filename)
+
+        upload_path = config['UPLOAD_FOLDER'] + secure_filename(form.csv_file.data.filename)
+        column_names = get_column_names(upload_path)
+        schema_definition = ", ".join([col_name + " STRING " for col_name in column_names])
+
+        
+
+        import boto3
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: import on top of module

File: superset/db_engine_specs.py [hunk 28]
```diff
@@ -686,6 +805,50 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
...
+        s3 = boto3.client('s3')
+        s3.upload_file(upload_path, 'airbnb-superset', os.path.join(upload_prefix, table_name, filename))
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: line too long

File: superset/db_engine_specs.py [hunk 29]
```diff
@@ -686,6 +805,50 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
...
+        sql = "CREATE EXTERNAL TABLE {} ( {} ) ROW FORMAT DELIMITED FIELDS TERMINATED BY ',' STORED AS TEXTFILE LOCATION '{}'"\
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: as from previous comment, use triple quoted strings and kwargs as opposed to args in `.format`. Checkout the magic convention `.format(**locals())` to avoid boilerplate

File: superset/views/core.py [hunk 30]
```diff
@@ -298,6 +304,94 @@ class DatabaseAsync(DatabaseView):
...
+    form = CsvToDatabaseForm
+    form_title = _('CSV to Database configuration')
+    add_columns = ['database', 'schema', 'table_name']
+
+    def form_get(self, form):
+        # pre process form
+        # default values
+        form.name.data = None
```
Comment (mistercrunch) [inline superset/views/core.py]: Also lots of boilerplate here.

File: superset/views/core.py [hunk 31]
```diff
@@ -298,6 +304,94 @@ class DatabaseAsync(DatabaseView):
...
+        form.if_exists.data = 'append'
+        form.index.data = None
+        form.index_label.data = None
+
+    def form_post(self, form):
+        def upload_file(csv_file):
+            if csv_file and csv_file.filename:
+                filename = secure_filename(csv_file.filename)
+                csv_file.save(os.path.join(config['UPLOAD_FOLDER'],
+                                                 filename))
```
Comment (mistercrunch) [inline superset/views/core.py]: indent?

File: superset/views/core.py [hunk 32]
```diff
@@ -298,6 +304,94 @@ class DatabaseAsync(DatabaseView):
...
+
+
+        form.names.data = form.names.data.split(",") if form.names.data else 0
+
+        #datasources = ConnectorRegistry.get_all_datasources(db.session)
+        filename = upload_file(form.csv_file.data)
+
+        table = SqlaTable(table_name=form.name.data)
+
+        database = db.session.query(models.Database).filter_by(sqlalchemy_uri=form.data.get('con')).one()
```
Comment (mistercrunch) [inline superset/views/core.py]: line too long

**Review by mistercrunch — CHANGES_REQUESTED** [review]
Gave up before getting to the end as the PR is still full of lint and some previous comments have not been addressed. Sharpen your linter!

File: superset/csvform.py [hunk 33]
```diff
@@ -0,0 +1,205 @@
...
+      logging.info("Exception retreiving dbs: {}".format(e))
+
+    con = SelectField(_('Database URI'),
+                     description=_('URI of database in which to '
+                                    'add above table.'),
+                     validators=[DataRequired()],
+                     choices=all_datasources
+                       )
+    if_exists = SelectField(_('Table Exists'),
+                            description=_('If table exists do one '
```
[the remark this answers is not in the document]
Comment (timifasubaa) [reply superset/csvform.py]: Fixed

File: superset/views/core.py (hunk 30 above)
Comment (timifasubaa) [reply superset/views/core.py]: These set the default values. I removed the ones that weren't changing anything.

File: superset/db_engine_specs.py (hunk 29 above)
Comment (timifasubaa) [reply superset/db_engine_specs.py]: Done

File: superset/db_engine_specs.py (hunk 28 above)
Comment (timifasubaa) [reply superset/db_engine_specs.py]: Done

File: superset/db_engine_specs.py (hunk 27 above)
Comment (timifasubaa) [reply superset/db_engine_specs.py]: Done

File: superset/db_engine_specs.py (hunk 26 above)
Comment (timifasubaa) [reply superset/db_engine_specs.py]: Done

File: superset/db_engine_specs.py (hunk 25 above)
Comment (timifasubaa) [reply superset/db_engine_specs.py]: Fixed!

File: superset/db_engine_specs.py (hunk 24 above)
Comment (timifasubaa) [reply superset/db_engine_specs.py]: Fixed

File: superset/csvform.py [hunk 34]
```diff
@@ -0,0 +1,205 @@
...
+                                       'uploaded to a database.'),
+                         validators=[FileRequired(),
+                                     FileAllowed(['csv'],
+                                                 _('CSV Files '
+                                                   'Only!'))])
+    all_datasources = []
+    try:
+      all_datasources = db.session.query(models.Database.sqlalchemy_uri, models.Database.database_name).all()
+    except Exception as e:
+      logging.info("Exception retreiving dbs: {}".format(e))
```
[the remark this answers is not in the document]
Comment (timifasubaa) [reply superset/csvform.py]: Because whenever I run it, it works but it breaks when the CI is running. The error it produces is: 
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such table: dbs [SQL: 'SELECT dbs.sqlalchemy_uri AS dbs_sqlalchemy_uri, dbs.database_name AS dbs_database_name \nFROM dbs']

It needs the dbs table that is not existent on the travis instance but the table will exist whenever someone runs superset.

File: superset/forms.py [hunk 35]
```diff
@@ -0,0 +1,191 @@
...
+        validators=[DataRequired()],
+        widget=BS3TextFieldWidget())
+    csv_file = FileField(
+        _('CSV File'),
+        description=_('Select a CSV file to be uploaded to a database.'),
+        validators=[
+            FileRequired(), FileAllowed(['csv'], _('CSV Files Only!'))])
+    all_datasources = []
+    try:
+        all_datasources = db.session\
```
Comment (mistercrunch) [inline superset/forms.py]: This should be in view scope, not in module scope. As is, it would be interpreted when the web server (re)starts

**Review by mistercrunch — CHANGES_REQUESTED** [review]

File: superset/forms.py (hunk 35 above)
Comment (timifasubaa) [reply superset/forms.py]: Fixed

File: superset/db_engine_specs.py [hunk 36]
```diff
@@ -74,6 +84,81 @@ def extra_table_metadata(cls, database, table_name, schema_name):
...
+        df_to_db_kwargs = {
+            'table': table,
+            'df': df,
+            'name': form.name.data,
+            'con': form.con.data,
+            'schema': form.schema.data,
+            'if_exists': form.if_exists.data,
+            'index': form.index.data,
+            'index_label': form.index_label.data,
+            'chunksize': 10000
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: NIT: we like trailing commas, we should see if pylint can require it

File: superset/db_engine_specs.py [hunk 37]
```diff
@@ -686,6 +771,58 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
         return BaseEngineSpec.fetch_result_sets(
             db, datasource_type, force=force)
 
+    @staticmethod
+    def upload_csv(form, table):
+        "uploads a csv file and creates a superset table in the right databse"
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: https://www.python.org/dev/peps/pep-0257/

File: superset/db_engine_specs.py [hunk 38]
```diff
@@ -686,6 +771,58 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
...
+            return False
+
+        upload_prefix = config["CSV_TO_HIVE_UPLOAD_DIRECTORY"]
+        dest_path = os.path.join(table_name, filename)
+
+        upload_path = config['UPLOAD_FOLDER'] + secure_filename(
+            form.csv_file.data.filename)
+        column_names = get_column_names(upload_path)
+        schema_definition =\
+            ", ".join([col_name + " STRING " for col_name in column_names])
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: NIT: s/col_name/s/g

File: superset/db_engine_specs.py [hunk 39]
```diff
@@ -686,6 +771,58 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
...
+
+        s3 = boto3.client('s3')
+        location =\
+            os.path.join("s3a://", bucket_path, upload_prefix, table_name)
+        s3.upload_file(
+            upload_path,
+            'airbnb-superset',
+            os.path.join(upload_prefix, table_name, filename))
+        sql = "CREATE EXTERNAL TABLE {table_name} ( {schema_definition} ) "\
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: NIT: 3rd comment about triple-quote strings

File: superset/translations/es/LC_MESSAGES/messages.po [hunk 40]
```diff
@@ -7,7 +7,7 @@ msgid ""
 msgstr ""
```
Comment (mistercrunch) [inline superset/translations/es/LC_MESSAGES/messages.po]: damn dude. these `.po` files are still in your PR

File: superset/views/core.py [hunk 41]
```diff
@@ -7,18 +7,18 @@
 from datetime import datetime, timedelta
 import json
 import logging
-import pandas as pd
 import pickle
 import re
 import time
 import traceback
-
+import os
```
Comment (mistercrunch) [inline superset/views/core.py]: NIT: when there are many imports in a module, standard lib import should be together in a group (empty line between import groups), and alpha-sorted within that group preferably 

organizing imports becomes more important as there are more imports in the module

File: superset/views/core.py [hunk 42]
```diff
@@ -298,6 +301,64 @@ class DatabaseAsync(DatabaseView):
...
+                .filter_by(sqlalchemy_uri=form.data.get('con')).one()
+
+            message = _('CSV file "{0}" uploaded to table "{1}" in '
+                        'database "{2}"'.format(form.csv_file.data.filename,
+                                                form.name.data,
+                                                db_name[0]))
+            flash(message, 'info')
+            return redirect('/tablemodelview/list/')
+
+
```
Comment (mistercrunch) [inline superset/views/core.py]: NIT: one empty line too many

File: superset/views/core.py [hunk 43]
```diff
@@ -7,18 +7,18 @@
...
-
+import os
+import pandas as pd
 import sqlalchemy as sqla
 
 from flask import (
     g, request, redirect, flash, Response, render_template, Markup,
-    url_for)
-from flask_appbuilder import expose
+    url_for, send_from_directory)
```
Comment (mistercrunch) [inline superset/views/core.py]: used?

File: superset/views/core.py [hunk 44]
```diff
@@ -298,6 +301,64 @@ class DatabaseAsync(DatabaseView):
...
+        form.skip_blank_lines.data = True
+        form.parse_dates.data = True
+        form.infer_datetime_format.data = True
+        form.dayfirst.data = False
+        form.decimal.data = '.'
+        form.error_bad_lines.data = False
+        form.if_exists.data = 'append'
+        all_datasources = []
+        all_datasources = db.session.query(
+            models.Database.sqlalchemy_uri, models.Database.database_name)\
```
Comment (mistercrunch) [inline superset/views/core.py]: NIT: line break would be much better after the coma

File: superset/views/core.py [hunk 45]
```diff
@@ -298,6 +301,64 @@ class DatabaseAsync(DatabaseView):
...
+
+
+    def form_post(self, form):
+        def upload_file(csv_file):
+            if csv_file and csv_file.filename:
+                filename = secure_filename(csv_file.filename)
+                csv_file.save(os.path.join(config['UPLOAD_FOLDER'], filename))
+                return filename
+
+        form.names.data = form.names.data.split(",") if form.names.data else 0
```
Comment (mistercrunch) [inline superset/views/core.py]: Seems like `form.names.data` should receive consistent types, not `list` or `Int` depending... Maybe `[]` or `None`?

File: superset/views/core.py [hunk 46]
```diff
@@ -298,6 +301,64 @@ class DatabaseAsync(DatabaseView):
...
+
+        database = db.session.query(models.Database)\
```
Comment (mistercrunch) [inline superset/views/core.py]: Made a comment on this PR about how we like the method chaining to be wrap across lines before.
database = (
    db.session.query(models.Database)
    .filter_by(sqlalchemy_uri=form.data.get('con'))
    .one()
)

File: superset/forms.py [hunk 47]
```diff
@@ -0,0 +1,183 @@
...
+from flask_appbuilder.fieldwidgets import BS3TextFieldWidget
+from flask_appbuilder.forms import DynamicForm
+from flask_wtf.file import FileField, FileAllowed, FileRequired
+from wtforms import (
+    BooleanField, SelectField, IntegerField, StringField)
+from wtforms.validators import DataRequired, Optional, NumberRange
+
+from superset import app, db
+import superset.models.core as models
+
```
Comment (mistercrunch) [inline superset/forms.py]: NIT: uneeded blank line

File: superset/db_engine_specs.py (hunk 36 above)
Comment (mistercrunch) [reply superset/db_engine_specs.py]: (maybe pylint does, and flake8 doesn't)

File: superset/views/core.py (hunk 41 above)
Comment (timifasubaa) [reply superset/views/core.py]: Fixed

File: superset/db_engine_specs.py (hunk 38 above)
Comment (timifasubaa) [reply superset/db_engine_specs.py]: Not sure what you mean here.

File: superset/db_engine_specs.py [hunk 48]
```diff
@@ -74,6 +84,82 @@ def extra_table_metadata(cls, database, table_name, schema_name):
...
+        def allowed_file(filename):
+            # Only allow specific file extensions as specified in the config
+            extension = os.path.splitext(filename)[1]
+            return extension and extension[1:] in config['ALLOWED_EXTENSIONS']
+
+        filename = secure_filename(form.csv_file.data.filename)
+        if not allowed_file(filename):
+            flash("Invalid file type selected.", 'alert')
+            return False
+        kwargs = {
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: I don't think I was super clear about the kwargs comment earlier, and the verbose boilerplate isn't that bad either (others may consider the verbose, metaprogramming-free option better). 

Knowing the form object has a data dict:
http://wtforms.simplecodes.com/docs/0.6.1/forms.html#wtforms.form.Form.data

You can do something like:
```
form_data = form.data
csv_to_df_fields = ['names', 'filepath_or_buffer', ...]
kwargs = {s: form_data.get(s) for s in csv_to_df_fields}
kwargs['chunksize'] = 10000
# other alterations as needed
df = BaseEngineSpec.csv_to_df(**kwargs)
```

File: superset/db_engine_specs.py [hunk 49]
```diff
@@ -74,6 +84,82 @@ def extra_table_metadata(cls, database, table_name, schema_name):
...
+            config['UPLOAD_FOLDER'] + kwargs['filepath_or_buffer']
+        kwargs['encoding'] = 'utf-8'
+        kwargs['iterator'] = True
+        chunks = pandas.read_csv(**kwargs)
+        df = pandas.DataFrame()
+        df = pandas.concat(chunk for chunk in chunks)
+        return df
+
+    @staticmethod
+    def df_to_db(**kwargs):
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: if you change the function signature to `def df_to_db(df, table, con, schema, **kwargs):` then you don't need the 5 following lines and avoid juggling with kwargs

File: superset/db_engine_specs.py [hunk 50]
```diff
@@ -74,6 +84,82 @@ def extra_table_metadata(cls, database, table_name, schema_name):
...
+        df = kwargs['df']
+        table = kwargs['table']
+        del kwargs['df']
+        del kwargs['table']
+        kwargs['con'] = create_engine(kwargs['con'], echo=False)
+        df.to_sql(**kwargs)
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: having taken stuff out of kwargs in the signature, now here you may to pass those things explicitely as in `df.to_sql(con=create_engine(con, schema=schema, echo=False), **kwargs)`

File: superset/views/core.py [hunk 51]
```diff
@@ -298,6 +299,65 @@ class DatabaseAsync(DatabaseView):
 appbuilder.add_view_no_menu(DatabaseAsync)
 
 
+class CsvToDatabaseView(SimpleFormView):
+    form = CsvToDatabaseForm
+    form_title = _('CSV to Database configuration')
+    add_columns = ['database', 'schema', 'table_name']
+
+    def form_get(self, form):
+        form.sep.data = ','
```
Comment (mistercrunch) [inline superset/views/core.py]: I think the normal way of setting defaults for wtforms is to do it when creating the form objects. As in:
`BooleanField('name', default=True, ...)`

File: superset/views/core.py [hunk 52]
```diff
@@ -298,6 +299,65 @@ class DatabaseAsync(DatabaseView):
...
+        form.mangle_dupe_cols.data = True
+        form.skipinitialspace.data = False
+        form.skip_blank_lines.data = True
+        form.parse_dates.data = True
+        form.infer_datetime_format.data = True
+        form.dayfirst.data = False
+        form.decimal.data = '.'
+        form.error_bad_lines.data = False
+        form.if_exists.data = 'append'
+        all_datasources = db.session.query(
```
Comment (mistercrunch) [inline superset/views/core.py]: this one may have to stay here since you don't want to do it in module scope...

**Review by mistercrunch — CHANGES_REQUESTED** [review]
We're starting to get somewhere, should we add some unit tests?

File: superset/db_engine_specs.py [hunk 53]
```diff
@@ -686,6 +772,56 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
...
+                return csv.reader(f).next()
+
+        table_name = form.name.data
+        filename = form.csv_file.data.filename
+
+        bucket_path = config["CSV_TO_HIVE_UPLOAD_BUCKET"]
+
+        if not bucket_path:
+            logging.info("No upload bucket specified")
+            flash(
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: i18n

Comment (timifasubaa) [reply superset/db_engine_specs.py]: Fixed

File: superset/views/core.py (hunk 52 above)
Comment (timifasubaa) [reply superset/views/core.py]: fixed
